File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        def twoSum(cur, target, start):
            seen = set()
            n = start
            while n < len(nums):
                comp = target-nums[n]
                if comp in seen:
                    res.append([cur, nums[n], comp])
                    while n+1 < len(nums) and nums[n] == nums[n+1]:
                        n += 1
                else:
                    seen.add(nums[n])
                # skip processed nums
                n += 1
                
                
        
        for i in range(len(nums)-2):
            if nums[i] > 0:
                break
            if (i > 0 and nums[i] != nums[i-1]) or i == 0:
                twoSum(nums[i], -nums[i], i+1)
        
        return res

File: turingplanet/hashtable/15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */
#include "lib/headers.h"

// @lc code=start
class Solution
{
  public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        int n = nums.size();

        for (int i = 0; i < n - 2; i++)
        {
            if (nums[i] > 0)
                break;  // not possible to find value
            if (i > 0 && nums[i] == nums[i - 1])
                continue;  // already processed

            int j = i + 1, k = n - 1;
            while (j < k)
            {
                long sum = static_cast<long>(nums[i]) + nums[j] + nums[k];
                if (sum == 0)
                {
                    res.push_back({nums[i], nums[j], nums[k]});
                    // shrink the search window
                    while (j < k && nums[j] == nums[j + 1])
                        j++;
                    while (j < k && nums[k] == nums[k - 1])
                        k--;
                    // move to next number
                    j++;
                    k--;
                }
                else if (sum < 0)
                {
                    // need a larger j
                    j++;
                }
                else
                {
                    // need a smaller k
                    k--;
                }
            }
        }
        return res;
    }
};
// @lc code=end
